join: be less magical and handle magic with less undefined behaviour

tonycoz opened the pull request:
Fixes #21458 

Also fixed a couple of other minor problems.

tonycoz committed:
join: tests from GH #21458

tonycoz committed:
join: add tests from my comments on #21484

tonycoz committed:
join: save the delimiter string before anything magical happens to it

This code had a few problems:

- changes to the content of delim from set or overload magic could
  result in the separator between elements changing during the
  process of the join.
- changes to the content of delim which allocated a new PVX
  resulted in access to freed memory
- changes to the flags of delim, the UTF-8 flag in particular, could
  result in an invalid joined string, either mojibake or an invalidly
  encoded upgraded string

To avoid that, we copy the separator, either into a local buffer
if it's large enough, or an allocated buffer, and save the flag we
use, to prevent changes to the delim SV from changing or invalidating
the delimpv value.

Fixes #21458 and some similar problems.

tonycoz committed:
do_join: handle more than 2G arguments

No tests since I don't have the memory.

tonycoz committed:
do_join: validate arguments before we start using them

leonerd commented on doop.c:
```diff
@@ -658,13 +665,30 @@ Magic and tainting are handled.
-    const char * const delims = SvPV_const(delim, delimlen);
-
-    PERL_ARGS_ASSERT_DO_JOIN;
+    const char * delimpv = SvPV_const(delim, delimlen);
```
Excellent - you renamed it `delimpv` :)

leonerd commented on t/op/join.t:
```diff
@@ -128,3 +128,84 @@ package o { use overload q|""| => sub { ${$_[0]}++ } }
+    my $expect = "a\x{100}x\x{100}b\n";
+    utf8::encode($expect);
+    fresh_perl_is(<<'CODE', $expect, {}, "modifications delim from magic should be ignored");
+my $n = 1;
```
I'm not quite sure what `$n` is doing in this test. It is only ever 1 and only ever used as `STR x $n`. It could probably be removed for simplicity, unless it is being used to defeat some const-folding..? But it doesn't appear so to me.

tonycoz replied:
I've added an explanation of why I'm doing the `x $n` in each case, and yes, it's being done to prevent constant folding, or at least prevent CoW between `$sep` and some SVOP/PADOP.

If you wondered about it then and explanation can save someone else from having to do code archaeology trying to figure out the intent of the text.

leonerd replied:
Ahah, yes that makes sense. A good comment for some otherwise-subtle logic.

tonycoz committed:
join tests: comments clarifying why we don't do simple assignment

I changed $n to 2 in each case as a defence against a reasonable
possible optimisation of $x x 1 sharing the PV with $x.

leonerd reviewed (approved):
LGTM
